ARROW-11759: [C++] Kernel to extract datetime components (year, month, day, etc) from timestamp type

rok opened the pull request:
This is to resolve [ARROW-11759](https://issues.apache.org/jira/browse/ARROW-11759).

rok commented:
Missing features at moment:
* isocalendar
* millisecond
* microsecond
* nanosecond

Different input / output types should also be handled.
Does this need to be timezone aware?

jorisvandenbossche commented:
> Does this need to be timezone aware?

Unfortunately yes .. But since that's a lot more complicated, and I don't think we already have tz-related functionality, for now those kernels could maybe just raise an error if the timestamp type has a `tz` and is thus not tz-naive?

rok commented:
It seems tests on windows are not able to download IANA timezone database or access it post-download (https://howardhinnant.github.io/date/tz.html#Installation). I'll solve it but meanwhile I I'd like to ask for review.

Open questions I have at the moment:
- Is it possible to match kernels to zoned timestamps when registering or do we have to keep matching at runtime [like so](https://github.com/apache/arrow/blob/2765cefc737ac8b5c364a20f8ff94fc158b91b37/cpp/src/arrow/compute/kernels/scalar_temporal.cc#L58).
- What would be a good array type for isodate?

rok commented:
I've adopted [some code](https://github.com/apache/arrow/blob/2765cefc737ac8b5c364a20f8ff94fc158b91b37/cpp/src/arrow/compute/kernels/scalar_temporal.cc#L197) from [iso_week.h](https://github.com/HowardHinnant/date/blob/master/include/date/iso_week.h#L1503). Would it make more sense to add iso_week.h as a vendored library instead? Should we add a note to our license? (iso_week.h is MIT licensed).

pitrou commented on cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:
```diff
@@ -0,0 +1,239 @@
+  CheckScalarUnary("nanosecond", time_points, nanosecond);
+}
+
+TEST(ScalarTemporalTest, TestTemporalComponentExtraction) {
```
Is this test different from the previous one?

pitrou commented on cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:
```diff
@@ -0,0 +1,239 @@
+  auto second = ArrayFromJSON(int64(), "[59, 23, 20, 20]");
+  auto millisecond = ArrayFromJSON(int64(), "[0, 0, 0, 0]");
+  auto microsecond = ArrayFromJSON(int64(), "[0, 0, 0, 0]");
+  auto nanosecond = ArrayFromJSON(int64(), "[0, 0, 0, 0]");
```
Hmm... so you never test with a non-zero number of milli/micro/nanoseconds?
Also, how about testing with null values as well?

pitrou commented on cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:
```diff
@@ -0,0 +1,239 @@
+  auto second = ArrayFromJSON(int64(), "[59, 23, 1, 20, 20]");
+  auto millisecond = ArrayFromJSON(int64(), "[0, 0, 0, 0, 0]");
+  auto microsecond = ArrayFromJSON(int64(), "[0, 0, 0, 0, 0]");
+  auto nanosecond = ArrayFromJSON(int64(), "[0, 0, 0, 0, 0]");
```
Is it desirable for all these values to be `int64`? A number of hours, minutes or seconds should fit in a `int8`, for example.

pitrou commented on cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:
```diff
@@ -0,0 +1,239 @@
+}
+
+TEST(ScalarTemporalTest, TestZonedTemporalComponentExtraction) {
+  std::string timezone = "Etc/GMT+2";
```
Again, is this testing something different compared to above?

pitrou commented on cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:
```diff
@@ -0,0 +1,239 @@
+          "3989-07-14T18:04:01","1900-01-01T01:59:20","2033-05-18T03:33:20"])";
+  auto time_points = ArrayFromJSON(timestamp(TimeUnit::SECOND, "Etc/GMT+2"), json);
+
+  auto year = ArrayFromJSON(int64(), "[1969, 2000, 3989, 1899, 2033]");
```
Hmm... can you explain the reasoning here? If the timestamp is situated in 1970, I would expect the year to be 1970.

rok replied:
Yeah, I suppose this is taking `UTC` year instead of `Etc/GMT+2` year. Will fix.

rok replied:
It turns out `Etc/GMT+X == UTC-X` so this was actually ok. Weird mapping.

pitrou commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -0,0 +1,632 @@
+using arrow_vendored::date::literals::thu;
+
+template <typename Duration>
+inline year_month_day ymd_caster_template(const int64_t data) {
```
Please follow the coding convention guidelines, i.e. `CamelCase` for functions.

rok replied:
Removed functions so this is probably fixed now :)

pitrou commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -0,0 +1,632 @@
+// ----------------------------------------------------------------------
+// Extract year from timestamp
+
+template <typename out_type>
```
As per the coding conventions, this should probably `OutType`.

rok replied:
Done.

pitrou commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -0,0 +1,632 @@
+    }
+    return Status::OK();
+  }
+};
```
You're repeating yourself a lot. You basically have specific extraction function (the "caster" functions) and a generic kernel structure that just applies the extraction function. It should be possible to minimize the amount of code copied and pasted in this file.

rok replied:
Indeed. If I "dry" this up do you feel the approach is ok?

rok replied:
I've abstracted the "caster maker" function. I'll try to do the same with the generic kernel.

pitrou reviewed (changes requested):
Thanks for the PR @rok. There's a bunch of things that look unexpected to me, see below.

jorisvandenbossche commented:
General comment: the PR is quite big, and the timezone-related logic further added complexity. So from a workflow perspective (to ensure reviewing / getting this merged is manageable), could it make sense to keep the timezone-related logic for a separate PR? (didn't check the code to see whether that's actually feasible)

rok commented:
> General comment: the PR is quite big, and the timezone-related logic further added complexity. So from a workflow perspective (to ensure reviewing / getting this merged is manageable), could it make sense to keep the timezone-related logic for a separate PR? (didn't check the code to see whether that's actually feasible)

Keeping tz-aware logic would be fairly easy to move to a separate PR. I'll move it out.

jorisvandenbossche commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -0,0 +1,348 @@
+// Based on
+// https://github.com/HowardHinnant/date/blob/6e921e1b1d21e84a5c82416ba7ecd98e33a436d0/include/date/iso_week.h#L1503
+template <typename Duration>
+struct Week {
```
We should probably be explicit here about it being the "ISO week" number (and not the "logical" week number), as it gives some surprising results around New Year (https://en.wikipedia.org/wiki/ISO_week_date#First_week)

(and if adding a ISO week, we should maybe also add the corresponding ISO year)

rok replied:
I'll do that.
By the way what would be a good data type for "isocalendar" output? It's a [tuple in Pandas](https://pandas.pydata.org/docs/reference/api/pandas.Timestamp.isocalendar.html).

jorisvandenbossche replied:
A struct? (with year, week, day fields)

jorisvandenbossche commented on cpp/src/arrow/compute/api_scalar.h:
```diff
@@ -450,5 +450,145 @@ ARROW_EXPORT
+/// \since 4.0.0
+/// \note API not yet finalized
+ARROW_EXPORT
+Result<Datum> Nanosecond(const Datum& values, ExecContext* ctx = NULLPTR);
```
Are this the total nanoseconds, or only the nanoseconds part after the microseconds? (i.e. are microseconds and nanoseconds additive components, or rather both representing the full fractional part of the second using a different resolution?)

jorisvandenbossche replied:
Seeing the tests, it's clearly the first (additive components), which I think is good. I am only wondering if we can make that clearer in the terminology of the docstring

rok replied:
Oh yeah indeed. This would behave in the latter way which would be surprising to users. Will fix.

jorisvandenbossche replied:
To be clear, judging the tests, the current implementation is not using the "total" nanoseconds right now, but only the part after the microseconds. Personally I think this is the useful behaviour

rok replied:
Do you know by hear what does pandas do? [Documentation](https://pandas.pydata.org/docs/reference/api/pandas.Series.dt.microsecond.html) is not clear.
Perhaps it would be good to then have a `subseconds` function that would return the total nanoseconds since midnight?

jorisvandenbossche replied:
Yeah, I was also looking at pandas to compare ;) And that was the reason I am asking.

So the situation is a bit complex there:

```
>>> ts = pd.Timestamp("2012-01-02 03:04:05.123456789")
>>> ts.second
5
```

There is a `microsecond` attribute which gives the total microseconds, but this is inherited from `datetime.datetime` (as this actually stores the sub-second part as microseconds):

```
>>> ts.microsecond
123456
```

and then there is also a `nanosecond` attribute which is only the part after the microseconds (so the "additive componenent", not total nanoseconds):
```
>>> ts.nanosecond
789
```


(the example in the pandas docs is indeed not good! As you can't tell which of the two cases it is because of the used value)

rok replied:
Well this is not ideal :).
I propose to implement this as `total units since midnight` and then add an exception in python for nanoseconds.

jorisvandenbossche replied:
We don't need to match exactly what pandas does to be clear. pyarrow.compute should just follow Arrow's behaviour.

So for Arrow the question is, what is most useful: the total units vs the component? (or do we need both?)

As comparison, the R lubridate package only has component getter functions up to `second()`, and not for `milli/micro/..second`, but the `second` function returns the fractional part of the second as well:

```R
> library(lubridate)
> x <- ymd_hms("2012-03-26 01:02:03.123456")
> second(x)
[1] 3.123456
```

rok replied:
How about for `2012-03-26 01:02:03.123456789`:
* subsecond -> 123456789
* second[int] -> 3
* second[float] -> 3.123456789
* microsecond -> 123
* millisecond -> 456
* nanosecond -> 789

This way we have minimal `postprocessing` in languages other than c++

rok commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -0,0 +1,509 @@
+
+template <typename Duration>
+struct ISOCalendar {
+  static Status Call(KernelContext* ctx, const Scalar& in, Scalar* out) {
+    using ScalarType = typename TypeTraits<Int64Type>::ScalarType;
+
+    const auto& in_val = internal::UnboxScalar<const TimestampType>::Unbox(in);
+    const auto dp = sys_days{floor<days>(Duration{in_val})};
+    const auto ymd = year_month_day(dp);
+    auto y = year_month_day{dp + days{3}}.year();
+    auto start = sys_days((y - years{1}) / dec / thu[last]) + (mon - thu);
+    if (dp < start) {
+      --y;
+      start = sys_days((y - years{1}) / dec / thu[last]) + (mon - thu);
+    }
+
+    std::vector<std::shared_ptr<Scalar>> values = {
+        std::make_shared<ScalarType>(
+            static_cast<int64_t>(static_cast<int32_t>(ymd.year()))),
+        std::make_shared<ScalarType>(
+            static_cast<int64_t>(trunc<weeks>(dp - start).count() + 1)),
+        std::make_shared<ScalarType>(static_cast<int64_t>(weekday(ymd).iso_encoding()))};
+    *checked_cast<StructScalar*>(out) =
+        StructScalar(std::move(values), iso_calendar_type);
+    return Status::OK();
+  }
```
@pitrou any idea why this scalar kernel is failing here? The test outputs and [CI log](https://github.com/apache/arrow/pull/10176/checks?check_run_id=2666780178#step:10:1299) is not very telling.

rok replied:
Got it. Kernel signature was set to only expect arrays and not scalars.

rok commented:
@jorisvandenbossche @pitrou - other then the iso_calendar I feel this is ready for another review round.

rok commented:
I fixed the iso_calendar issue so I'd ask for another round of review please :)

pitrou commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -0,0 +1,614 @@
+      uint16(), int16(), uint32(), int32(), uint64(), int64()};
+  static std::vector<std::shared_ptr<DataType>> kUnsignedIntegerTypes32 = {
+      uint32(), int32(), uint64(), int64()};
+  static std::vector<std::shared_ptr<DataType>> kSignedIntegerTypes = {int32(), int64()};
```
I don't understand why you're having all these types. The input type is timestamp, right?

rok replied:
[See](https://github.com/apache/arrow/pull/10176#discussion_r643454925)

rok commented on cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:
```diff
@@ -0,0 +1,107 @@
+  ASSERT_RAISES(Invalid, Millisecond(timestamps));
+  ASSERT_RAISES(Invalid, Microsecond(timestamps));
+  ASSERT_RAISES(Invalid, Nanosecond(timestamps));
+  ASSERT_RAISES(Invalid, Subsecond(timestamps));
```
Wouldn't ignoring be a bit unexpected? It would be relatively easy to add timezone processing now - [see here](https://github.com/apache/arrow/commit/36abbcb512b2b45c8b5615fe29074be2afbd01d8).

pitrou replied:
What is the expected semantics? It seems to me that the results would simply returned in the local timezone, i.e. if I have a timestamp for '2000-01-01 00:00:00' in the UTC+1 timezone, then the year is 2000... in the UTC+1 timezone.

If users want to convert to UTC first, perhaps we want a separate kernel for that, but I see no reason to do that conversion in the datetime component extraction kernels.

For example in Python:
```python
>>> t = dateutil.tz.gettz('Europe/Paris')
>>> dt = datetime.datetime(2000,1,1,0,0,0, tzinfo=t)
>>> dt
datetime.datetime(2000, 1, 1, 0, 0, tzinfo=tzfile('/usr/share/zoneinfo/Europe/Paris'))
>>> dt.year
2000
>>> t.utcoffset(dt)
datetime.timedelta(seconds=3600)
>>> dt.utctimetuple().tm_year
1999
```

rok replied:
Arrow timestamp is UTC unix time with optional timezone information (if I understand correctly).
So if I want to know e.g. hour in my timezone I need to convert to my timezone at some point.
Timezone conversion kernel is an option, but then timestamp becomes local time instead of unix time?

pitrou replied:
Ah... It seems you're right indeed. 
```
          /// * If the time zone is set to a valid value, values can be displayed as
          ///   "localized" to that time zone, even though the underlying 64-bit
          ///   integers are identical to the same data stored in UTC. Converting
          ///   between time zones is a metadata-only operation and does not change the
          ///   underlying values
```

I will ask on the ML for clarification.

pitrou replied:
In any case, this is mostly a documentation issue. If the timestamp value is always in UTC, then "year" and friends will return the year expressed in the UTC timezone. I don't think there is any reason to forbid the operation, just document it properly.

rok replied:
ML discussion on this would be good indeed.
About raising when timezone is present - as a user I would be slightly surprised but could live with it.
@joris [previously suggested raising](https://github.com/apache/arrow/pull/10176#issuecomment-828254868).

I'm indifferent on this but would like add timezone processing ([like so](https://github.com/rok/arrow/commit/36abbcb512b2b45c8b5615fe29074be2afbd01d8)).

jorisvandenbossche replied:
Indeed, the storage model of timezone-aware data is different compared to stdlib python, we are always storing the UTC timestamp (and not the localized timestamp).

> In any case, this is mostly a documentation issue. If the timestamp value is always in UTC, then "year" and friends will return the year expressed in the UTC timezone. I don't think there is any reason to forbid the operation, just document it properly.

I personally disagree with this. If you have a timezone-aware timestamp, as a user I would expect to get the result in *local*time. If I want to have the UTC version, I can easily first convert to UTC (which is a no-op). 
For "year" it most of the time doesn't change much (only the hours around New Year), but if I have a timestamp "2021-06-01 09:00:00+02:00" (CEST), and I ask for "hour", I expect to get 9 and not 7.

pitrou replied:
Hmm, that makes sense.

rok replied:
Let's leave it as is then. I'll open a followup JIRA to add timezone processing.

rok commented on cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:
```diff
@@ -0,0 +1,107 @@
+  CheckScalarUnary("millisecond", unit, times, int64(), millisecond);
+  CheckScalarUnary("microsecond", unit, times, int64(), microsecond);
+  CheckScalarUnary("nanosecond", unit, times, int64(), nanosecond);
+  CheckScalarUnary("subsecond", unit, times, int64(), subsecond);
```
We'll probably need to calculate these to mimic pandas and R. We can calculate them here or later. [See discussion.](https://github.com/apache/arrow/pull/10176#discussion_r633564079)

rok commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -0,0 +1,614 @@
+using arrow_vendored::date::literals::mon;
+using arrow_vendored::date::literals::thu;
+
+// Based on ScalarUnaryNotNullStateful. Adds timezone awareness.
```
Done.

rok commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -0,0 +1,614 @@
+struct Year {
+  template <typename T>
+  static T Call(KernelContext*, int64_t arg, Status*) {
+    return static_cast<T>(static_cast<const int32_t>(
```
date.h won't allow casting to types other than int32 and throws at compile-time:
```
../src/arrow/compute/kernels/scalar_temporal.cc:144:78: error: invalid static_cast from type 'arrow_vendored::date::year' to type 'const int64_t' {aka 'const long int'}
  144 |         year_month_day(floor<days>(sys_time<Duration>(Duration{arg}))).year()));
```
I'll take a look if there's a recommended way to deal with this.

pitrou replied:
Ah, ok. You can just add a comment then.

rok replied:
Similar discussion here for [weekaday](https://github.com/HowardHinnant/date/issues/391). Didn't find a nice way to fix it.
This happens in many locations as date.h is quite strict with types. Will add comments.

rok commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -0,0 +1,614 @@
+struct DayOfYear {
+  template <typename T>
+  static T Call(KernelContext*, int64_t arg, Status*) {
+    const auto sd = sys_days{floor<days>(Duration{arg})};
```
`sys_days` is shorthand for `sys_time<days>`. I've made everything `sys_time<days>` now.

rok commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -0,0 +1,614 @@
+    const auto& out_type = TypeTraits<OutType>::type_singleton();
+
+    const std::string timezone =
+        std::static_pointer_cast<const TimestampType>(in.type)->timezone();
```
Done.

rok commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -0,0 +1,614 @@
+      if (dp < start) {
+        --y;
+        start = sys_days((y - years{1}) / dec / thu[last]) + (mon - thu);
+      }
```
Done.

rok commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -0,0 +1,614 @@
+
+    std::unique_ptr<ArrayBuilder> array_builder;
+    RETURN_NOT_OK(MakeBuilder(ctx->memory_pool(), iso_calendar_type, &array_builder));
+    StructBuilder* struct_builder = checked_cast<StructBuilder*>(array_builder.get());
```
Done.

rok commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -0,0 +1,614 @@
+    field_builders.reserve(3);
+    for (int i = 0; i < 3; i++) {
+      field_builders.push_back(
+          checked_cast<BuilderType*>(struct_builder->field_builder(i)));
```
Done.

rok commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -0,0 +1,614 @@
+template <template <typename... Args> class KernelGenerator,
+          template <typename... Args> class Op, typename Duration>
+ArrayKernelExec ExecFromOp(detail::GetTypeId get_id) {
+  switch (get_id.id) {
```
Inputs are timestamps but I'd like to control type of output (as proposed [here](https://github.com/apache/arrow/pull/10176#discussion_r629457595)). I'm not sure output type of a kernel can be matched?

pitrou replied:
I don't think there's a point in trying to control the output type. Also this will generate a lot of additional code.

rok replied:
Wouldn't control of output type be useful sometimes?
Will remove.

rok replied:
Removed.

rok commented:
Thanks for the review @pitrou! I've addressed some of the comments and I'll try to finish the rest today.

rok commented on cpp/src/arrow/compute/api_scalar.h:
```diff
@@ -462,5 +462,177 @@ ARROW_EXPORT
+ARROW_EXPORT
+Result<Datum> Year(const Datum& values, ExecContext* ctx = NULLPTR);
+
+/// \brief Month returns month value for each element of `values`
```
Done.

rok commented on cpp/src/arrow/compute/api_scalar.h:
```diff
@@ -462,5 +462,177 @@ ARROW_EXPORT
+ARROW_EXPORT
+Result<Datum> Month(const Datum& values, ExecContext* ctx = NULLPTR);
+
+/// \brief Day returns day value for each element of `values`
```
Done.

rok commented on cpp/src/arrow/compute/api_scalar.h:
```diff
@@ -462,5 +462,177 @@ ARROW_EXPORT
+/// \note API not yet finalized
+ARROW_EXPORT Result<Datum> DayOfWeek(const Datum& values, ExecContext* ctx = NULLPTR);
+
+/// \brief DayOfYear returns day of year value for each element of `values`
```
Done.

rok commented on cpp/src/arrow/compute/api_scalar.h:
```diff
@@ -462,5 +462,177 @@ ARROW_EXPORT
+/// \note API not yet finalized
+ARROW_EXPORT Result<Datum> DayOfYear(const Datum& values, ExecContext* ctx = NULLPTR);
+
+/// \brief ISOYear returns ISO year value for each element of `values`
```
Done.

rok commented on cpp/src/arrow/compute/api_scalar.h:
```diff
@@ -462,5 +462,177 @@ ARROW_EXPORT
+ARROW_EXPORT
+Result<Datum> ISOYear(const Datum& values, ExecContext* ctx = NULLPTR);
+
+/// \brief ISOWeek returns ISO week of year value for each element of `values`
```
Done.

rok commented on cpp/src/arrow/compute/api_scalar.h:
```diff
@@ -462,5 +462,177 @@ ARROW_EXPORT
+/// \note API not yet finalized
+ARROW_EXPORT Result<Datum> ISOCalendar(const Datum& values, ExecContext* ctx = NULLPTR);
+
+/// \brief Quarter returns quarter of year value for each element of `values`
```
Done. :)

rok commented on cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:
```diff
@@ -0,0 +1,107 @@
+  auto iso_year = "[1970, 2000, 1899, 2033, null]";
+  auto iso_week = "[1, 9, 52, 20, null]";
+  auto iso_calendar = ArrayFromJSON(iso_calendar_type,
+                                    R"([{"iso_year": 1970, "iso_week": 1, "weekday": 4},
```
`day_of_week`

rok commented on docs/source/cpp/compute.rst:
```diff
@@ -637,6 +637,54 @@ String extraction
+
+* \(1) The ISO 8601 definition for week 01 is the week with the first Thursday
+  of the Gregorian year (i.e. of January) in it.
+  .. _Wikipedia ISO Week date: https://en.wikipedia.org/wiki/ISO_week_date#First_week
```
Changed.

rok commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -0,0 +1,614 @@
+};
+
+// ----------------------------------------------------------------------
+// Extract day of quarter from timestamp
```
Done.

rok commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -0,0 +1,614 @@
+const FunctionDoc millisecond_doc{"Extract millisecond values", "", {"values"}};
+const FunctionDoc microsecond_doc{"Extract microsecond values", "", {"values"}};
+const FunctionDoc nanosecond_doc{"Extract nanosecond values", "", {"values"}};
+const FunctionDoc subsecond_doc{"Extract subsecond values", "", {"values"}};
```
Done.
I feel very awkward about my English right now.

rok commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -0,0 +1,614 @@
+  template <typename T>
+  static T Call(KernelContext*, int64_t arg, Status*) {
+    return static_cast<T>(static_cast<const int32_t>(
+        year_month_day{sys_days{floor<days>(Duration{arg})} + days{3}}.year()));
```
Added comment above.

rok commented on docs/source/cpp/compute.rst:
```diff
@@ -637,6 +637,54 @@ String extraction
++--------------------+------------+-------------------+-----------------+--------+
+| day                | Unary      | Temporal          | Numeric         |        |
++--------------------+------------+-------------------+-----------------+--------+
+| day_of_week        | Unary      | Temporal          | Numeric         |        |
```
Done.

rok commented on docs/source/cpp/compute.rst:
```diff
@@ -637,6 +637,54 @@ String extraction
+  of the Gregorian year (i.e. of January) in it.
+  .. _Wikipedia ISO Week date: https://en.wikipedia.org/wiki/ISO_week_date#First_week
+
+* \(2) Struct with fields ISO year, ISO week number, and weekday.
```
Done.

rok commented on docs/source/cpp/compute.rst:
```diff
@@ -637,6 +637,54 @@ String extraction
++--------------------+------------+-------------------+-----------------+--------+
+| Function name      | Arity      | Input types       | Output type     | Notes  |
++====================+============+===================+=================+========+
+| year               | Unary      | Temporal          | Numeric         |        |
```
Done.

rok commented:
@pitrou It seems kernels do not match on units of timestamps?
[This test fails](https://github.com/apache/arrow/pull/10176/commits/efd6b9e746dadaaaa6d1cc69614e6145b30b8037#diff-19248dfd2de25c17622290aae9c78024ba474de2913e44bc8cf0dce0bfe7dcdaR107) for every unit except for nanosecond even though we [register kernels for all units](https://github.com/apache/arrow/pull/10176/commits/efd6b9e746dadaaaa6d1cc69614e6145b30b8037#diff-53850537a2f3f7977473b30241fdd3ddb35aa97269284c06471d9b0c599e2c59R415).
Is this expected? We can get timestamp units at runtime but matching would be preferable I think.

Other than that I still need to [add a couple of test times](https://github.com/apache/arrow/pull/10176#discussion_r643061679).

jorisvandenbossche commented on cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:
```diff
@@ -0,0 +1,151 @@
+  auto quarter = "[1, 1, 1, 2, null]";
+  auto hour = "[0, 23, 0, 3, null]";
+  auto minute = "[0, 23, 59, 33, null]";
+  auto second = "[59.123456789, 23.999999999, 20.001001001, 20.0, null]";
```
I would have expected an integer here, which is the number for second in standard ISO string representation (just like the hour and minute etc)

(so at least this should be documented more clearly, if we keep this behaviour)

rok replied:
Switching to integer.

jorisvandenbossche commented on cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:
```diff
@@ -0,0 +1,151 @@
+  auto millisecond = "[123, 999, 1, 0, null]";
+  auto microsecond = "[456, 999, 1, 0, null]";
+  auto nanosecond = "[789, 999, 1, 0, null]";
+  auto subsecond = "[123456789, 999999999, 1001001, 0, null]";
```
I would have expected a float here, but of course an integer is possible as well (see my question above on the docstring of this function for asking to clarify it, but so my inline suggestion there was assuming it would be a float).

Now, if we return an integer, what is this number exactly? Does it then depend on the unit of the Timestamp, or are it always nanoseconds (even when the unit is eg milliseconds) 
Returning it as a float would avoid the ambiguity (but of course for certain applications an integer might be more useful).

rok replied:
I've changed this to float now as it is more intuitive indeed. Definition is now seconds since last full second.

If it was an integer as it was before the definition would be: total nanoseconds since last full second.

jorisvandenbossche commented on cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:
```diff
@@ -0,0 +1,151 @@
+  auto second = "[59, 23, 20, 20, null]";
+
+  //  for (auto u : internal::AllTimeUnits()) {
+  for (auto u : {TimeUnit::NANO}) {
```
This is not yet testing the different units?

rok replied:
As stated above:
> This test fails for every unit except for nanosecond even though we [register kernels for all units](https://github.com/apache/arrow/pull/10176/commits/efd6b9e746dadaaaa6d1cc69614e6145b30b8037#diff-53850537a2f3f7977473b30241fdd3ddb35aa97269284c06471d9b0c599e2c59R415).
Is this expected? We can get timestamp units at runtime but matching would be preferable I think.

jorisvandenbossche replied:
Ah, sorry, missed that comment.

jorisvandenbossche commented on cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:
```diff
@@ -0,0 +1,151 @@
+  auto quarter = "[1, 1, 1, 2, null]";
+  auto hour = "[0, 23, 0, 3, null]";
+  auto minute = "[0, 23, 59, 33, null]";
+  auto second = "[59, 23, 20, 20, null]";
```
Can you also test the other sub-second componenents here? (even for a unit of "second", those components should be defined I think?)

rok replied:
I've currently not added these because ArrayFromJSON fails when timestamps are more precise than their units. E.g.:
```
CheckScalarUnary("millisecond", ArrayFromJSON(timestamp(TimeUnit::SECOND), R"(["1970-01-01T00:00:59.123"])"), ArrayFromJSON(float64(), "[0]"));
```
Fails with:
```
-- Arrow Fatal Error --
Invalid: couldn't parse timestamp from 1970-01-01T00:00:59.123
```

I'll think about how to do this nicely.

jorisvandenbossche replied:
The original timestamps don't need to be more detailed (to start with), you can leave those as is. But then I would still expect the subcomponents to work and return 0 ?

rok replied:
Sure, I'll add a block like:
```
    CheckScalarUnary("millisecond", unit, times, int64(), zeros);
    CheckScalarUnary("microsecond", unit, times, int64(), zeros);
    CheckScalarUnary("nanosecond", unit, times, int64(), zeros);
    CheckScalarUnary("subsecond", unit, times, float64(), zeros);
```

Where:
```
auto zeros = "[0, 0, 0, 0, null]";
```

jorisvandenbossche reviewed (commented):
Thanks for the updates, here are some more comments

rok commented on cpp/src/arrow/compute/api_scalar.h:
```diff
@@ -481,5 +481,184 @@ ARROW_EXPORT
+/// \note API not yet finalized
+ARROW_EXPORT Result<Datum> ISOWeek(const Datum& values, ExecContext* ctx = NULLPTR);
+
+/// \brief ISOCalendar returns a (ISO year, ISO week, weekday) struct for each element of
```
`Day of week`?

jorisvandenbossche commented:
To recap some spread inline comments (original discussion at https://github.com/apache/arrow/pull/10176#discussion_r633497393) about the sub-second / fractional components. 

Assume the timestamp `2012-03-26 01:02:03.123456789`.

Currently, I think the PR does:

* second -> 3.123456789 (float)
* subsecond -> 123456789 (int)

Based on a naive reading of the docstrings and my biased expectations, I would have thought they return:

* second -> 3 (int)
* subsecond -> 0.123456789 (float)


Now, in our previous discussion you proposed to implement `subsecond` as "total units since last second", so that indeed matches with what is in the PR now. But, that also means that the interpretation of this value critically depends on the unit of the timestamp type. While a float value would not have that problem.

In addition, I personally think it would be useful to have some form of `second` that is just the integer the non-fractional component in " 01:02:03.123...".

---

In addition, we also have the other components (for which eg Antoine questioned whether they are needed):

* microsecond -> 123
* millisecond -> 456
* nanosecond -> 789

rok commented on docs/source/cpp/compute.rst:
```diff
@@ -637,6 +637,55 @@ String extraction
+  starts with the first ISO week.
+  See `ISO 8601 week date definition`_ for more details.
+* \(3) Output is a ``{"iso_year": output type, "iso_week": output type, "day_of_week":  output type}`` Struct.
+.. _ISO 8601 week date definition: https://en.wikipedia.org/wiki/ISO_week_date#First_week
```
Sounds like a good idea!

rok commented on cpp/src/arrow/compute/api_scalar.h:
```diff
@@ -481,5 +481,184 @@ ARROW_EXPORT
+ARROW_EXPORT
+Result<Datum> Millisecond(const Datum& values, ExecContext* ctx = NULLPTR);
+
+/// \brief Microsecond returns microseconds value for each element of `values`
```
Done.

jorisvandenbossche commented:
There seems to be something wrong with the ISO calendar (year) implementation. I was playing a bit with this branch, and seeing:

```python
In [25]: arr = pa.array([pd.Timestamp("2012-01-01 01:02:03")], type=pa.timestamp("ns"))

In [26]: pc.iso_calendar(arr)
Out[26]: 
<pyarrow.lib.StructArray object at 0x7f5bbf3639a0>
-- is_valid: all not null
-- child 0 type: int64
  [
    2012            #  <--------------  this should be 2011 (since the week is 52, it falls in the "previous" year)
  ]
-- child 1 type: int64
  [
    52
  ]
-- child 2 type: int64
  [
    7
  ]
```

Also in general, for non-nanosecond resolutions, you get wrong results for all kernels. It seems that it is interpreting the non-nanosecond epoch as nanosecond epoch. That's related to your previous comment that tests for other units are still failing?

rok commented on docs/source/cpp/compute.rst:
```diff
@@ -637,6 +637,55 @@ String extraction
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+These functions extract datetime components (year, month, day, etc) from timestamp type.
+Note: timezone information is currently ignored if present.
```
Updated.

rok commented:
> There seems to be something wrong with the ISO calendar (year) implementation.

Will take a look. Still need to add test cases too.

> Also in general, for non-nanosecond resolutions, you get wrong results for all kernels. It seems that it is interpreting the non-nanosecond epoch as nanosecond epoch. That's related to your previous comment that tests for other units are still failing?

Yeah. It seems that currently all timestamps are interpreted as if having nanosecond unit. I suspect it's a kernel matching problem.

rok commented:
Thanks for the review @jorisvandenbossche! I think I've addressed the second/subsecond issue and docs. (Sorry for amend commits).

I'll now look into:
* iso_year/iso_week/day_of_week tests
* all timestamps being interpreted as having nanosecond unit

jorisvandenbossche commented:
> (Sorry for amend commits).

Don't be sorry for amending commits. Personally that's actually my strong preference that you amend commits during the review process, then we can see what changed since a previous review.

jorisvandenbossche commented:
I did a quick naive check of the performance of this branch, and comparing the simple components, they are consistently 6 to 25x faster compared to the pandas implementation (6x for the year/month/day, 25-30x for hour/minute/second). That aspect seems to be good! 
(I don't know how good the pandas implementation is, so it might not be the most relevant comparison (comparing with eg clickhouse might be more interesting), but it at least says something)

rok commented:
> I did a quick naive check of the performance of this branch, and comparing the simple components, they are consistently 6 to 25x faster compared to the pandas implementation (6x for the year/month/day, 25-30x for hour/minute/second). That aspect seems to be good!
> (I don't know how good the pandas implementation is, so it might not be the most relevant comparison (comparing with eg clickhouse might be more interesting), but it at least says something)

Nice! Thanks for measuring that. That's a dramatic improvement!
Maybe we could still optimize somewhat but that would require a bit of a study of `date.h` and benchmarking as changes are tried.

I've fixed the `iso_year` function and added some tests [borrowed from pandas](https://github.com/pandas-dev/pandas/blob/059c8bac51e47d6eaaa3e36d6a293a22312925e6/pandas/tests/tslibs/test_ccalendar.py#L39).


I think the last thing left is the matching to proper time unit.

rok commented:
@jorisvandenbossche - I found the cause of the unit issue - it was [ the missing`break`](https://github.com/apache/arrow/pull/10176/files#diff-53850537a2f3f7977473b30241fdd3ddb35aa97269284c06471d9b0c599e2c59R425)s in a switch statement. Typical. :)

I think all comments were addressed now but more eyballs == better code :).

rok commented:
CI issues don't seem related.
I've also created a [followup Jira](https://issues.apache.org/jira/browse/ARROW-12980) for the timezone aware component extraction.

rok commented:
> CI issues don't seem related.
> I've also created a [followup Jira](https://issues.apache.org/jira/browse/ARROW-12980) for the timezone aware component extraction.

jorisvandenbossche commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -0,0 +1,626 @@
+}
+
+template <template <typename...> class Op>
+std::shared_ptr<ScalarFunction> MakeSimpleTemporal(std::string name,
```
```suggestion
std::shared_ptr<ScalarFunction> MakeStructTemporal(std::string name,
```

? (the kernel returning a struct type doesn't seem "simpler")

jorisvandenbossche commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -0,0 +1,626 @@
+
+const FunctionDoc iso_year_doc{
+    "Extract ISO year number",
+    ("First week of an ISO year has the majority (4 or more) of it's days in January."
```
```suggestion
    ("First week of an ISO year has the majority (4 or more) of its days in January."
```

jorisvandenbossche commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -0,0 +1,626 @@
+
+const FunctionDoc iso_week_doc{
+    "Extract ISO week of year number",
+    ("First ISO week has the majority (4 or more) of it's days in January.\n"
```
```suggestion
    ("First ISO week has the majority (4 or more) of its days in January.\n"
```

jorisvandenbossche commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -0,0 +1,626 @@
+    {"values"}};
+
+const FunctionDoc hour_doc{
+    "Extract hour values",
```
```suggestion
    "Extract hour value",
```

(above you also use singular form for eg "day number")

(and the same for the ones below)

jorisvandenbossche commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -0,0 +1,626 @@
+    {"values"}};
+
+const FunctionDoc millisecond_doc{
+    "Extract millisecond values",
```
Can you add here the clarification about it being the number of milliseconds since the last full second, as you did for the C++ functions? (and same for microsecond/nanosecond below)

rok replied:
Done!

jorisvandenbossche commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -0,0 +1,626 @@
+    {"values"}};
+
+const FunctionDoc subsecond_doc{
+    "Extract subsecond values",
```
Here also best to add the clarification of what subsecond is from the `Subsecond` docstring

rok replied:
Done!

jorisvandenbossche commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -0,0 +1,626 @@
+  }
+  return {static_cast<int64_t>(static_cast<int32_t>(y)),
+          static_cast<int64_t>(trunc<weeks>(t - start).count() + 1),
+          static_cast<int64_t>(weekday(ymd).iso_encoding())};
```
In the DayOfWeek implementation you subtract 1 (to have 0-6), while here not. I assume that's the ISO definition? If so, we should clearly document that difference in the ISOCalendar documentation (since now it seems to imply that the "day_of_week" field is the same as the "day_of_week" kernel. Maybe also prepend it with "iso_" ?

rok replied:
Added to docs and prepended.

jorisvandenbossche commented on docs/source/cpp/compute.rst:
```diff
@@ -665,6 +665,57 @@ String component extraction
++--------------------+------------+-------------------+---------------+--------+
+| iso_week           | Unary      | Temporal          | Int64         | \(2)   |
++--------------------+------------+-------------------+---------------+--------+
+| iso_calendar       | Unary      | Temporal          | Scalar Struct | \(3)   |
```
```suggestion
| iso_calendar       | Unary      | Temporal          | Struct        | \(3)   |
```

(not sure what you mean with the "scalar" struct here?)

jorisvandenbossche reviewed (commented):
Thanks for the updates! Did a final pass through the docs. 

Also, it would maybe be good to add a test for a timestamp that doesn't fit into the nanosecond range?

rok commented:
Thanks for the comments and suggestions @jorisvandenbossche! :)

pitrou commented on docs/source/cpp/compute.rst:
```diff
@@ -665,6 +665,57 @@ String component extraction
   e.g. 'letter' and 'digit' for the regular expression
   ``(?P<letter>[ab])(?P<digit>\\d)``.
 
+Temporal component extraction
```
Nit: can you move these below other string-related sections?

rok replied:
Moved. Please check if the location is ok.

pitrou commented:
Thank you for this PR @rok ! I believe this is ready for merging. I'll just wait for CI.

jorisvandenbossche commented on docs/source/cpp/compute.rst:
```diff
@@ -860,6 +860,58 @@ null input value is converted into a null output value.
+* \(2) First ISO week has the majority (4 or more) of it's days in January. ISO year
+  starts with the first ISO week.
+  See `ISO 8601 week date definition`_ for more details.
+* \(3) Output is a ``{"iso_year": output type, "iso_week": output type, "day_of_week":  output type}`` Struct.
```
```suggestion
* \(3) Output is a ``{"iso_year": output type, "iso_week": output type, "iso_day_of_week":  output type}`` Struct.
```

Just noted one last detail, but @rok let's leave this for your next PR to fix this, so we can let CI continue here and get this merged ;)

rok commented:
Just noticed I missed this:
> @jorisvandenbossche
> Also, it would maybe be good to add a test for a timestamp that doesn't fit into the nanosecond range?

Do you mind if I move it to [ARROW-12980](https://github.com/apache/arrow/pull/10457)?

jorisvandenbossche commented:
Cool, this is a nice start for datetime kernels! 
@thisisnic opened a JIRA for expsosing those in R as well (https://issues.apache.org/jira/browse/ARROW-13022)

> what do you mean by "doesn't fit into the nanosecond range"?

I meant a date that falls outside the *date* range of 1677-09-21 - 2262-04-11 (the range that ns resolution can cover).

rok commented:
Thanks @jorisvandenbossche @pitrou for the help and feedback! I'm very glad this is merged! :)
